[motion-path] Make <ray-size> optional in ray()

nt1m opened the pull request:
#### 1a13191fa42070e3a194e34eccb71d2be2a3d2c2
<pre>
[motion-path] Make &lt;ray-size&gt; optional in ray()
<a href="https://bugs.webkit.org/show_bug.cgi?id=258110">https://bugs.webkit.org/show_bug.cgi?id=258110</a>
rdar://110818689

Reviewed by Antti Koivisto and Darin Adler.

In the updated motion path spec, the size in the offset path ray() function is now optional. If not specified, the size
parameter defaults to `closest-side`.

Also implement logic to omit `closest-side` when serializing, since we should always serialize to the shortest form.

Spec: <a href="https://drafts.fxtf.org/motion-1/#ray-function">https://drafts.fxtf.org/motion-1/#ray-function</a>

Fix pre-existing some unefficient idioms as well:
- Switch to consumeIdentRaw for parsing isContaining flag to avoid creating a CSSValue
- Make CSSRayValue store a CSSValueID for size instead of a CSSValue
- Make CSSRayValue use makeString for serialization

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-supports-calc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-parsing-valid-expected.txt:
* Source/WebCore/css/CSSRayValue.cpp:
(WebCore::CSSRayValue::customCSSText const):
(WebCore::CSSRayValue::equals const):
* Source/WebCore/css/CSSRayValue.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForPathOperation):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeRayShape):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertPathOperation):

Canonical link: <a href="https://commits.webkit.org/265200@main">https://commits.webkit.org/265200@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/27e0345b6dbe94842ff4f0d2b6b8138acdaae1fe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10225 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10463 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10725 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11871 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/9855 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10237 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12457 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10415 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12796 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10383 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11127 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8620 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12257 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8432 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9261 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/16550 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9531 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9413 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12660 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9869 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9028 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2449 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13276 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9698 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSRayValue.cpp:
```diff
@@ -35,10 +35,10 @@ String CSSRayValue::customCSSText() const
+    builder.append("ray(", m_angle->cssText());
+
+    if (m_size->valueID() != CSSValueClosestSide)
+        builder.append(" ", m_size->cssText());
```
This should use `‘ ‘` since appending a character is more efficient than appending a string even when the string is known at compile time in our current implementation. Also it _could_ be refactored to use a single call to `makeString` instead. That is not particularly challenging. but note that we might switch back to `StringBuilder` if we can start fixing these to not use intermediate strings but appending the text of sub-values with append functions

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -6952,10 +6953,14 @@ static RefPtr<CSSRayValue> consumeRayShape(CSSParserTokenRange& range, const CSS
+    // <angle> must be present.
+    if (!angle)
         return nullptr;
 
```
Seems like `CSSRayValue` should store a `CSSValueID`, not a `CSSValue`, for `ray-size`. More efficient in memory use and also for almost every operation we would do with it, like serialization.

also, for parsing the `isContaining` boolean we should use `consumeIdentRaw` so we don’t create and then immediately destroy a `CSSPrimitiveValue` as part of parsing.

darinadler reviewed (approved):
OK as is. But noticed some unnecessarily inefficient idioms in the code.

nt1m committed:
[motion-path] Make <ray-size> optional in ray()
https://bugs.webkit.org/show_bug.cgi?id=258110
rdar://110818689

Reviewed by Antti Koivisto and Darin Adler.

In the updated motion path spec, the size in the offset path ray() function is now optional. If not specified, the size
parameter defaults to `closest-side`.

Also implement logic to omit `closest-side` when serializing, since we should always serialize to the shortest form.

Spec: https://drafts.fxtf.org/motion-1/#ray-function

Fix pre-existing some unefficient idioms as well:
- Switch to consumeIdentRaw for parsing isContaining flag to avoid creating a CSSValue
- Make CSSRayValue store a CSSValueID for size instead of a CSSValue
- Make CSSRayValue use makeString for serialization

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-supports-calc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-parsing-valid-expected.txt:
* Source/WebCore/css/CSSRayValue.cpp:
(WebCore::CSSRayValue::customCSSText const):
(WebCore::CSSRayValue::equals const):
* Source/WebCore/css/CSSRayValue.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForPathOperation):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeRayShape):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertPathOperation):

Canonical link: https://commits.webkit.org/265200@main
